File: method/process/cpu/mrr.py
import sys

import numpy as np
from numpy.linalg import norm
from mpi4py import MPI

if __name__ == "__main__":
    sys.path.append('../../../../')
    from krylov.method.common import getConditionParams
    from krylov.method.process.cpu.common import init, start, end, matvec, vecvec 


def mrr(A, b, epsilon, T=np.float64):
    # 初期化
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    x, b_norm, N, max_iter, residual, num_of_solution_updates = init(A, b, T)

    # 初期残差
    r = b - matvec(A, x, comm)  # dot
    residual[0] = norm(r) / b_norm

    # 初期反復
    if rank == 0:
        start_time = start(method_name='MrR')
    Ar = matvec(A, r, comm)  # dot
    zeta = vecvec(r, Ar, comm) / vecvec(Ar, Ar, comm)  # dot
    y = zeta * Ar
    z = -zeta * r
    r -= y
    x -= z
    num_of_solution_updates[1] = 1

    # 反復計算
    for i in range(1, max_iter):
        # 収束判定
        residual[i] = norm(r) / b_norm
        isConverged = np.array([residual[i] < epsilon], dtype=bool)
        comm.Bcast(isConverged, root=0)
        if isConverged:
            break

        # 解の更新
        Ar = matvec(A, r, comm)  # dot
        nu = vecvec(y, Ar, comm)  # dot
        gamma = nu / vecvec(y, y, comm)  # dot
        s = Ar - gamma * y
        zeta = vecvec(r, s, comm) / vecvec(s, s, comm)  # dot
        eta = -zeta * gamma
        y = eta * y + zeta * Ar
        z = eta * z - zeta * r
        r -= y
        x -= z
        num_of_solution_updates[i] = i + 1
    else:
        isConverged = False

    num_of_iter = i + 1
    residual_index = i
    if rank == 0:
        end(start_time, isConverged, num_of_iter, residual, residual_index)

    
if __name__ == "__main__":
    A, b, epsilon, k, T = getConditionParams('condition.json')
    mrr(A, b, epsilon, T)

File: cpp/mrr.cpp
#include "MyBlas.h"

using namespace std;

template<typename T>
vector<T> mrr(const Matrix &A, const vector<T> &b, const double epsilon);

int main() {
  Matrix A = {
		{5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0}
	};
	vector<double> b = {3.0, 1.0, 4.0, 0.0, 5.0, -1.0, 6.0, -2.0, 7.0, -15.0};
  const double epsilon = 1e-8;

  auto x = mrr(A, b, epsilon);
  for(auto e: x) cout << e << endl;
}

template<typename T>
vector<T> mrr(const Matrix &A, const vector<T> &b, const double epsilon){
  const ul N = A.size();
  const ul max_iter = N*2;
  const double b_norm = vecvec(b, b);
	vector<T> x(N, 0.0);

  // 初期残差
	vector<T> r = b - A*x;
  double residual = vecvec(r, r) / b_norm;

  // 初期反復
  vector<T> Ar = matvec(A, r);
  double zeta = vecvec(r, Ar) / vecvec(Ar, Ar);
  vector<T> y = zeta * Ar;
  vector<T> z = -zeta * r;
  r = r - y;
  x = x - z;
  ul i = 1;

  // 反復計算
  while (i < max_iter){
    // 収束判定
    residual = vecvec(r, r) / b_norm;
    if (residual < epsilon) {
      break;
    }

    // 解の更新
    Ar = matvec(A, r);
    double mu = vecvec(y, y);
    double nu = vecvec(y, Ar);
    double gamma = nu / mu;
    vector<T> s = Ar - gamma * y;
    double rs = vecvec(r, s);
    double ss = vecvec(s, s);
    zeta = rs / ss;
    double eta = -zeta * gamma;
    y = eta * y + zeta * Ar;
    z = eta * z - zeta * r;
    r = r - y;
    x = x - z;
    i++;
  }

  return x;
}
